posthog/api/dashboards/dashboard_tiles.py

```diff
@@ -0,0 +1,165 @@
+from typing import Any, Callable, Dict, List, Optional
+
+import structlog
+from rest_framework import serializers, status, viewsets
+from rest_framework.decorators import action
+from rest_framework.exceptions import PermissionDenied, ValidationError
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from posthog.api.forbid_destroy_model import ForbidDestroyModel
+from posthog.api.routing import StructuredViewSetMixin
+from posthog.models import Dashboard, DashboardTile, Insight, User
+from posthog.models.activity_logging.activity_log import Change, Detail, log_activity
+from posthog.permissions import ProjectMembershipNecessaryPermissions, TeamMemberAccessPermission
+
+logger = structlog.get_logger(__name__)
+
+
+class SimpleDashboardTileSerializer(serializers.Serializer):
+    insight_id: serializers.IntegerField = serializers.IntegerField(required=True)
+    dashboard_id: serializers.IntegerField = serializers.IntegerField(required=True)
+
+    def validate_insight_id(self, value: int) -> int:
+        team_id = self.context["team_id"]
+        insight = Insight.objects.get(id=value)
+
+        if insight.team_id != team_id:
+            raise PermissionDenied("You cannot access this insight")
+
+        if insight.deleted:
+            raise ValidationError("You cannot use deleted insights in a dashboard tile")
+
+        return value
+
+    def validate_dashboard_id(self, value: int) -> int:
+        team_id = self.context["team_id"]
+        dashboard = Dashboard.objects.get(id=value)
+
+        if dashboard.team_id != team_id:
+            raise PermissionDenied("You cannot access this dashboard")
+
+        if dashboard.deleted:
+            raise ValidationError("You add tiles to deleted dashboards")
+
+        if (
+            dashboard.get_effective_privilege_level(self.context["request"].user.id)
+            == Dashboard.PrivilegeLevel.CAN_VIEW
+        ):
+            raise PermissionDenied(f"You don't have permission to add insights to dashboard: {dashboard.id}")
+
+        return value
+
+    def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> DashboardTile:
+        insight = Insight.objects.get(id=validated_data["insight_id"])
+        dashboard = Dashboard.objects.get(id=validated_data["dashboard_id"])
+
+        tile, created = DashboardTile.objects.get_or_create(dashboard=dashboard, insight=insight)
+        if not created and tile.deleted:
+            tile.deleted = False
+            tile.save()
+
+        return tile
+
+
+class DashboardTileViewSet(
+    StructuredViewSetMixin,
+    ForbidDestroyModel,
+    viewsets.GenericViewSet,
+):
+    queryset = DashboardTile.objects.all()
+    serializer_class = SimpleDashboardTileSerializer
+    filter_rewrite_rules = {"team_id": "dashboard__team_id"}
```

Comment:
We're digging ourselves deep with the ORM around dashboards in a way that's IMO stupidly hard to debug.

Q: What have we messed up? Should tiles have a team_id? Some other data is messed up?

Reply:
Specifically for `filter_rewrite_rules = {"team_id": "dashboard__team_id"}` this is the fifth usage of the `filter_rewrite_rules` capability of `StructuredViewSetMixin`. It feels redundant to me to track team_id on a tile. The tile is explicitly associated with the dashboard and can only ever have the dashboard's team id

More generally. Yep, the through model handling in the Django ORM seems lacking to me. Here the problem is that you can't filter the relation. We have Dashboard <--> Tile <--> (Insight|Text) and can soft delete any of the three components of the relation. There's no way to say that the `dashboard.insights` relation (or the reverse) is only where the through model is not deleted

Or that if the relation is changed the `tile` shouldn't be recreated (losing the data we're storing on the tile). Which is what has been happening when posting to insight API.

Ultimately the current tight relationship between dashboards and insights is being progressively relaxed. So that a dashboard has tiles. And those tiles might have insights.